vshard/router/init.lua

```diff
@@ -1136,33 +1171,37 @@ end
 --
 local function router_map_callrw(router, func, args, opts)
     local replicasets_to_map, err, err_id, map, rid
-    local timeout, do_return_raw, bucket_ids, plain_bucket_ids, grouped_args
+    local mode, timeout, do_return_raw, bucket_ids, plain_bucket_ids,
+        grouped_args
     if opts then
+        mode, err = router_set_map_callrw_mode(opts.mode, opts.bucket_ids)
+        if err then
+            return nil, err
+        end
         timeout = opts.timeout or consts.CALL_TIMEOUT_MIN
         do_return_raw = opts.return_raw
         bucket_ids = opts.bucket_ids
         plain_bucket_ids = util.table_is_numeric(bucket_ids) and bucket_ids or
             util.table_keys(bucket_ids)
     else
         timeout = consts.CALL_TIMEOUT_MIN
+        mode = MAP_CALLRW_FULL
     end
-    if plain_bucket_ids then
+    if mode == MAP_CALLRW_PARTIAL then
         timeout, err, err_id, rid, replicasets_to_map =
             router_ref_storage_by_buckets(router, plain_bucket_ids, timeout)
-        -- Grouped arguments are only possible with partial Map-Reduce.
-        if timeout then
-            grouped_args = router_group_map_callrw_args(
-                router, plain_bucket_ids, bucket_ids)
-        end
     else
         timeout, err, err_id, rid, replicasets_to_map =
             router_ref_storage_all(router, timeout)
     end
     if timeout then
+        if plain_bucket_ids then
```

Comment:
This won't work unfortunately. Previously the `router_group_map_callrw_args` was invoked right after `router_ref_storage_by_buckets`, now it can be invoked also after `router_ref_storage_all`. The first one finds the buckets in the cluster, updates the router map cache and only because of that we could use the `router_group_map_callrw_args` function. But the `router_ref_storage_all` doesn't do that.

Currently, the newly added mode (mode = 'full' + split args) doesn't work, when the cache on a router is broken, the arguments will be sent to the wrong instances. Let's won't forget to add the test case for that